rust/kernel/src/lib.rs

```diff
@@ -56,3 +59,24 @@ fn panic(_info: &PanicInfo) -> ! {
 
 #[global_allocator]
 static ALLOCATOR: allocator::KernelAllocator = allocator::KernelAllocator;
+
+/// Attempts to allocate memory for the given value using the global allocator. On success, the
+/// value is moved into it and returned to the caller wrapped in a Box.
+pub fn try_alloc<T>(value: T) -> KernelResult<Box<T>> {
+    let layout = core::alloc::Layout::for_value(&value);
+    let rptr = unsafe { alloc::alloc::alloc(layout) };
+    if rptr.is_null() {
+        return Err(Error::ENOMEM);
+    }
+    let ptr = rptr as *mut T;
+    unsafe {
+        ptr.write(value);
+        Ok(Box::from_raw(ptr))
+    }
+}
```

Comment:
It seems like the main thing this is doing is handling allocation failures. Since this is really orthagonal to what the patch is about (misc devices!), can we just use regular `Box::new()`?

Reply:
It is in the same PR, but it's in its own commit though.

Reply:
Right, it has its own commit precisely because it is not related to miscdevice. It is in the PR because miscdevice uses it.

As for using `Box::new()`, it doesn't allow us to gracefully deal with allocation failures. In fact, this is a wider discussion I think we all need to have: is there a way to handle failures in functions like Box::new() that I'm not aware of? If not, I think we should consider banning such usages from kernel code.

Reply:
No, it doesn't. I think there's some ongoing work in rust upstream on this issue. What do you think about moving this to it's own PR? I'd like to be able to handle them distinctly.

Reply:
> is there a way to handle failures in functions like Box::new() that I'm not aware of

There is a bunch of things to discuss regarding allocation in general (e.g. different allocations functions for `GFP_*` flags, allocators for data structures perhaps, etc.). For the moment we are just using the standard `alloc` + the global allocator feature rustc, but that could change.

Reply:
These are two small functions already in their own commit, that's why I thought they didn't need a whole PR of their own.

But I wouldn't mind moving to a different PR if you feel strongly about it.

Reply:
I'd find it helpful, if you wouldn't mind.

Reply:
Sure, I created #45.

Reply:
> There is a bunch of things to discuss regarding allocation in general (e.g. different allocations functions for `GFP_*` flags, allocators for data structures perhaps, etc.). For the moment we are just using the standard `alloc` + the global allocator feature rustc, but that could change.

What is the right forum to have discussions like these?

Reply:
The mailing list (I put it in the `README.md` but perhaps it is not visible enough?).

Reply:
And let me add you to the GitHub organization, too (allows you to re-run the CI etc.).

Reply:
> The mailing list (I put it in the `README.md` but perhaps it is not visible enough?).

No, I think it's fine, I should've paid more attention.

> And let me add you to the GitHub organization, too (allows you to re-run the CI etc.).

Thanks for that!